src/history.js

```diff
@@ -74,76 +81,46 @@ History.prototype = {
     return this.head
   },
 
-  invalidate () {
-    this.focus = null
-
-    this.invoke('unarchive', null)
-
-    this.reconcile()
-  },
-
   reconcile (action) {
-    // No need to run this function if there are no active repos
-    if (this.repos.length <= 0) {
-      return false
-    }
-
-    this.invoke('rollback', null)
-
-    this.rollforward()
-
-    this.invoke('release', action)
-  },
+    let focus = action
 
-  rollforward () {
-    let action = this.focus ? this.focus.next : this.root
-    let cacheable = true
+    while (focus) {
+      this.invoke('reconcile', focus)
 
-    while (action && action.parent !== this.head) {
-      if (!action.disabled) {
-        this.invoke('reconcile', action)
-      }
-
-      // Adjust the focus to the youngest disposable action. No need to
-      // rollforward through every single action all the time.
-      if (cacheable && action.disposable) {
-        this.focus = action
-        this.invoke('cache', this.archive())
+      if (focus === this.head) {
+        break
       } else {
-        cacheable = false
+        focus = focus.next
       }
-
-      action = action.next
     }
+
+    this.archive()
+
+    this.invoke('release', action)
   },
 
   archive () {
-    let shouldArchive = this.size > this.limit
+    let action = this.root
 
-    // Is the cache pointed at the base? If so, that means we need
-    // to purge the base.
-    if (shouldArchive) {
+    while (this.size > this.limit && action.disposable) {
       this.size -= 1
 
-      if (this.size <= 0) {
-        this.root = this.head = this.focus = null
-      } else {
-        this.root = this.root.next
-        this.root.parent = null
+      if (action.parent) {
+        this.invoke('clean', action.parent)
```

Comment:
I think there's a next step here where we don't dispatch a `clean` command until we've hit the new root, but I couldn't quite figure it out.